File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
  bool checkRow(vector<vector<char>> &board, int row) {
    unordered_set<char> set;
    for (int i = 0; i < 9; ++i) {
      if (board[row][i] != '.') {
        if (set.count(board[row][i])) {
          return false;
        }
        set.insert(board[row][i]);
      }
    }
    return true;
  }
  bool checkCol(vector<vector<char>> &board, int col) {
    unordered_set<char> set;
    for (int i = 0; i < 9; ++i) {
      if (board[i][col] != '.') {
        if (set.count(board[i][col])) {
          return false;
        }
        set.insert(board[i][col]);
      }
    }
    return true;
  }
  bool checkBox(vector<vector<char>> &board, int row, int col) {
    unordered_set<char> set;
    for (int i = row; i < row + 3; ++i) {
      for (int j = col; j < col + 3; ++j) {
        if (board[i][j] != '.') {
          if (set.count(board[i][j])) {
            return false;
          }
          set.insert(board[i][j]);
        }
      }
    }
    return true;
  }

 public:
  bool isValidSudoku(vector<vector<char>> &board) {
    bool output = true;
    for (int i = 0; i < 9; i++) {
      output = checkRow(board, i);
      if (!output) {
        return output;
      }
    }
    for (int i = 0; i < 9; i++) {
      output = checkCol(board, i);
      if (!output) {
        return output;
      }
    }
    for (int i = 0; i < 9; i += 3) {
      for (int j = 0; j < 9; j += 3) {
        output = checkBox(board, i, j);
        if (!output) {
          return output;
        }
      }
    }
    return true;
  }
};

File: Python/0036-valid-sudoku.py
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        def checkCols(board: List[List[str]], row: int) -> bool:
            boardSet = set()
            for i in range(9):
                if board[row][i] != '.':
                    if board[row][i] in boardSet:
                        return False
                    boardSet.add(board[row][i])
            return True

        def checkRows(board: List[List[str]], col: int) -> bool:
            boardSet = set()
            for i in range(9):
                if board[i][col] != '.':
                    if board[i][col] in boardSet:
                        return False
                    boardSet.add(board[i][col])
            return True

        def checkBox(board: List[List[str]], row: int, col: int) -> bool:
            boardSet = set()
            for i in range(row, row + 3):
                for j in range(col, col + 3):
                    if board[i][j] != '.':
                        if board[i][j] in boardSet:
                            return False
                    boardSet.add(board[i][j])
            return True

        for i in range(9):
            if not checkCols(board, i):
                return False
            if not checkRows(board, i):
                return False

        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                if not checkBox(board, i, j):
                    return False
        return True